File: knapsack/knapsack.cpp
/*
Rating: ~ 4.8 / 10
Link: https://open.kattis.com/problems/knapsack
*/

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

void fast() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
}

int main() {
    fast();
    int c,n;
    while (scanf("%d %d", &c,&n) == 2)
    {
        vector<int> weights, values;
        int cn = n;
        while (cn--) {
            int v,w;
            scanf("%d %d", &v, &w);
            values.push_back(v);
            weights.push_back(w);
        }
        vector<vector<int>> memo;
        vector<vector<int>> path_memo;
        memo.assign(n, vector<int>(c+1, -1));
        path_memo.assign(n, vector<int>(c+1, 0));
        // function<pair<int,int>(int,int,int)>
        function<pair<int,int>(int,int,int)> dp = [&](int id, int remainingW, int path) -> pair<int,int> {
            if ((id == n) || (remainingW == 0)) return make_pair(0, path);
            int &ans = memo[id][remainingW];
            int &path_ans = path_memo[id][remainingW];
            if (ans != -1) return make_pair(ans, path); //dp step!
            if (weights[id] > remainingW) {
                // we can't take, so we move on
                auto resp = dp(id + 1, remainingW, path); 
                ans = resp.first;
                path_ans = resp.second;
                return resp;
            }
            auto resp1 = dp(id + 1, remainingW, path); // we don't take
            auto resp2 = dp(id+1, remainingW - weights[id], path ^ (1 << id)); // we take
            resp2.first += values[id];
            if (resp1 > resp2) {
                ans = resp1.first;
                path_ans = resp1.second;
                return resp1;
            }
            else {
                ans = resp2.first;
                path_ans = resp2.second;
                return resp2;
            }
        };
        dp(0, c, 0);
        int path = path_memo[0][c];
        // cout << path << "\t" << memo[0][c] << endl;
        cout << memo[0][c] << endl;
        // cout << n << c << endl;
        vector<int> res;
        // for (auto line: memo) {
        //     for (auto v: line) {
        //         cout << v << " ";
        //     }
        //     cout << endl;
        // }
        // for (auto line: path_memo) {
        //     for (auto v: line) {
        //         cout << v << " ";
        //     }
        //     cout << endl;
        // }
        for (int i = 0; path; i++) {
            int t = 1 << i;
            if (path & t) {
                res.push_back(i);
                path ^= t;
            }
        }
        cout << res.size() << endl;
        for (int node : res) {
            cout << node << " ";
        }
        cout << endl;

    };
    return 0;
}

File: knapsack/knapsack.py
from sys import stdin

def main():
    while True:
        line = stdin.readline()
        if not line:
            break
        values = []
        weights = []
        c,n = map(int,line.split())
        for _ in range(n):
            v,w = map(int,stdin.readline().split())
            values.append(v)
            weights.append(w)
        memo = [[-1]*(c+1) for _ in range(n+1)]
        path_memo = [[0]*(c+1) for _ in range(n+1)]
        def dp(id, remainingW, path):
            if (id == n) or (remainingW == 0):
                return 0, path
            if memo[id][remainingW] != -1:
                return memo[id][remainingW], path_memo[id][remainingW]
            if weights[id] > remainingW:
                memo[id][remainingW], path_memo[id][remainingW] = dp(id+1, remainingW, path) # don't take id
                return memo[id][remainingW], path_memo[id][remainingW]
            pos1, path1 = dp(id+1, remainingW, path)
            pos2, path2 = dp(id+1, remainingW-weights[id], path | (1<<id))
            pos2 += values[id] 
            ans = (pos1, path1) if pos1 > pos2 else (pos2, path2)
            memo[id][remainingW] = ans[0]
            path_memo[id][remainingW] = ans[1]
            return memo[id][remainingW], path_memo[id][remainingW]
        dp(0, c, 0)
        path = path_memo[0][-1]
        res = []
        i = 0
        while path:
            t = 1 << i
            if path & t:
                res.append(str(i))
                path ^= t
            i += 1
        print(len(res), ' '.join(res), sep='\n')
if __name__ == '__main__':
    main()
